### PR TITLE
Basic strings

### LINKED ISSUE #383: Add log statements

### PR DESCRIPTION
**Pull request by vezenovm**
# Related issue(s)

Resolves #383 

The issue linked is to add log statements. Although not directly related, strings are being added to ultimately enable logging. 

# Description

This PR adds a basic string implementation. To correctly resolve string size we follow the same resolution strategy as arrays and under the hood we utilize Rust's String type and convert to a list of `u8` integers. 

## Summary of changes

I added a `String` type to various type enums such as `InputValue`, `UnresolvedType`, `Type` in the HIR, `Literal` in the monomorphisation AST, etc. Strings are allowed to be passed into the ABI and compared against each other. 

However, in this current implementation strings do not fully mock the functionality of arrays. This means we cannot replace individual elements of arrays, but a mutable string is possible by replacing the entire string value. 

I debated whether or not to implement a string type directly or rather create a char type and then enable arrays to handle this char type. This felt like it would be a little bit more complex as we would then have to extend arrays to support string literals, and char literals would have to be added as well. To start, I went with a separate string type as it seems highly unlikely that many people are going to be doing string manipulation in Noir and the main goal of adding strings was to enable logging which will just require string literals. I could make this change right now, or I could open an issue for later to add chars and enable formatting/manipulation of strings. 

## Dependency additions / changes

(If applicable.)

## Test additions / changes

I added a basic test checking whether or not we can pass string data to the ABI and compare it against string literals inside of Noir.

# Checklist

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` with default settings.
- [X] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this PR to the issue(s) that it resolves.
- [X] I have reviewed the changes on GitHub, line by line.
- [X] I have ensured all changes are covered in the description.

# Additional context

(If applicable.)

### CODE REVIEW

File: crates/noirc_frontend/src/hir/type_check/expr.rs
Comment (jfecher) [inline crates/noirc_frontend/src/hir/type_check/expr.rs:719]: This is a bug fixed in the numeric generics PR. Since these are types we should unify them instead of checking `!=`. Reason being that unification will correctly handle type variables.

File: crates/noirc_abi/src/input_parser/toml.rs
Comment (vezenovm) [reply crates/noirc_abi/src/input_parser/toml.rs:63]: We don't necessarily want a string, we may want a field element from the toml string. Previously we only let you input fields so it would be an error if the string could not be parsed as a field element.  I return `InputValue::Field(...)` if a field element is parsed as we want users to be able to still pass in hex strings to represent their field element inputs. However, I see how this is confusing. I'm going to look at moving this logic into the `parse_str` method and just returning an `InputValue` from it directly if this would make it more clear. I will add some comments as well.

File: crates/nargo/tests/test_data/strings/src/main.nr
Comment (vezenovm) [reply crates/nargo/tests/test_data/strings/src/main.nr]: I tried to have the syntax be `str<N>` but I kept running into this build error: 
```
error: Unexpected end of input, expected one of >, [, binary operator, cast, field access
  ┌─ /Users/maximvezenov/Documents/dev/noir/crates/nargo/tests/test_data/strings/src/main.nr:8:1
  │
8 │ }
  │ -
error: aborting due to 1 previous errors
```
I ended up going with this syntax for now `str[N]`. Do you have an idea of what might be causing this? The parser is the exact same except the tokens delimiting the expression for string size were `Token::Less` and `Token::Greater`.

File: crates/nargo/src/cli/prove_cmd.rs
Comment (vezenovm) [reply crates/nargo/src/cli/prove_cmd.rs]: Removed

File: crates/noirc_abi/src/input_parser/toml.rs
Comment (vezenovm) [reply crates/noirc_abi/src/input_parser/toml.rs]: I left a clarifying comment and renamed `parse_str` to `try_str_to_field`. Let me know what you think. I felt that checking for whether the string is field element first made sense as we do want an `InputValue::Field(...)` if the user passed a hex string as input to the circuit. Then if we cannot parse a field element we assume that the user wants to input a string.

File: crates/noirc_abi/src/lib.rs
Comment (vezenovm) [reply crates/noirc_abi/src/lib.rs:255]: The field element is going to convert to an array of bytes (size 32 for aztec_backend field), so I added an assert that the remaining bytes are all zero

File: crates/noirc_frontend/src/hir/type_check/expr.rs
Comment (vezenovm) [reply crates/noirc_frontend/src/hir/type_check/expr.rs:719]: Made the switch. I wasn't sure if I should put the change from your numeric generics PR in this one, but I put it in anyway as it is a small change.

File: crates/noirc_evaluator/src/ssa/code_gen.rs
Comment (vezenovm) [reply crates/noirc_evaluator/src/ssa/code_gen.rs:526]: Done

File: crates/noirc_abi/src/input_parser/toml.rs
Comment (jfecher) [inline crates/noirc_abi/src/input_parser/toml.rs:62]: Is this what we want for strings though? Wouldn't this prevent a user from passing in the string `"0x41"` by translating it to `"A"` instead?

File: crates/noirc_abi/src/lib.rs
Comment (jfecher) [reply crates/noirc_abi/src/lib.rs:188]: nitpicking now but `.bytes().map(...)` would be a bit more clear

File: crates/noirc_evaluator/src/ssa/code_gen.rs
Comment (jfecher) [reply crates/noirc_evaluator/src/ssa/code_gen.rs:525]: Don't need `into_iter()` here, `bytes` is already an iterator

File: crates/noirc_frontend/src/hir/type_check/expr.rs
Comment (jfecher) [inline crates/noirc_frontend/src/hir/type_check/expr.rs:723]: The error message here should mention strings instead of arrays

File: crates/noirc_frontend/src/parser/parser.rs
Comment (jfecher) [inline crates/noirc_frontend/src/parser/parser.rs:483]: Ah, here was your issue. You were using `expr_parser` for arbitrary expressions rather than the standard generics or type parser.
So for syntax like `str<N> ...` the expr_parser saw `N > ...` as an expression, and indeed distinguishing between `>` as an operator and as an end-type-bracket is ambiguous. The numeric generics PR doesn't actually allow parsing arbitrary expressions yet - only on array types which use the square brackets as well.
I was wondering whether to make it a separate PR but this can be fixed if we add a separate parser to only parse a subset of expressions valid in a type context. Notably those excluding the use of `>`.

File: crates/nargo/tests/test_data/strings/src/main.nr
Comment (guipublic) [reply crates/nargo/tests/test_data/strings/src/main.nr]: I don't like the `str[N]` syntax, as it look like we have an array of strings.
So I would go for `str<N>`, or only `char[N]` until we get the `str<N>` working.
Also, why not using `string` rather than `str`?

Comment (jfecher) [reply crates/nargo/tests/test_data/strings/src/main.nr]: We can change to `str<N>` once #620 is merged since that provides support for parsing numeric generics within regular brackets. Previously it was an ambiguous parse as we tried to parse arbitrary expressions where the terminating `>` is also a valid operator.
I'm guessing `str` was chosen over `string` here since rust also uses the `str` keyword for its string reference type `&str`.

File: crates/noirc_abi/src/input_parser/toml.rs
Comment (vezenovm) [reply crates/noirc_abi/src/input_parser/toml.rs:62]: Not entirely I guess as users cannot pass any utf8 string as input to a circuit. It would prevent a user from passing in the string "0x41" because it will be converted to a field element by the input parser, rather than translating it to "A". They can technically still have "0x41" as input but it must be a Field.

I mainly went with this method because of this comment `We still want developers to be able to pass in hex strings for FieldElement inputs.` Otherwise, we would have to add some more details to our input parser. We do compile our circuit before reading the inputs from file. I could add some logic where I pass along the ABI and then parse accordingly.

File: crates/noirc_frontend/src/hir/type_check/expr.rs
Comment (vezenovm) [reply crates/noirc_frontend/src/hir/type_check/expr.rs:740]: Done

File: crates/noirc_abi/src/input_parser/toml.rs
Comment (jfecher) [reply crates/noirc_abi/src/input_parser/toml.rs:62]: We'd still like that ability, but I think this should be type directed so that we can have both. It would be very confusing for users if they pass in something within quotes to a string field and the value of that string was not exactly what they wrote within the quotes.

File: crates/noirc_abi/src/lib.rs
Comment (jfecher) [inline crates/noirc_abi/src/lib.rs:46]: Did these serialization changes leak in from another PR?

**Review by jfecher — COMMENTED** [review]
Looks good. I take it it is currently blocked until #620 is merged

File: crates/noirc_abi/src/input_parser/toml.rs
Comment (vezenovm) [reply crates/noirc_abi/src/input_parser/toml.rs:62]: Fair point. I will look into passing the Abi when we parse the toml so that we can differentiate between parsing into a field element or a string

Comment (vezenovm) [reply crates/noirc_abi/src/input_parser/toml.rs]: I have now made it now so that the parsing for `TomlType::String` is directed by the abi types. I also added some extra inputs to my `strings` test using a string input and a field input for `0x41`.

File: crates/noirc_abi/src/input_parser/toml.rs
Comment (TomAFrench) [inline crates/noirc_abi/src/input_parser/toml.rs:20]: Passing a `BTreeMap<String, AbiType>` seems more natural here. We can get rid of the index counting and it'll be easier to reconcile with #677 (as then we can just pass the `AbiType`).

Comment (vezenovm) [reply crates/noirc_abi/src/input_parser/toml.rs]: Good note. I went with the `Vec<AbiType>` originally as the parameter name key in the toml map does not distinguish between structs and I made an assumption that this method would be simpler. However, reconstructing the Abi into a `BTreeMap<String, AbiType>` mocking the toml map as you said is definitely cleaner. I have changed this to use a `BTreeMap` now

File: crates/noirc_evaluator/src/ssa/code_gen.rs
Comment (jfecher) [inline crates/noirc_evaluator/src/ssa/code_gen.rs:579]: `.map(...).collect::<Vec<_>>()` can be a `vecmap`

File: crates/noirc_frontend/src/parser/parser.rs
Comment (jfecher) [inline crates/noirc_frontend/src/parser/parser.rs:580]: We can remove these comments

**Review by jfecher — COMMENTED** [review]
Looks good, just some minor cleanup with removing unneeded comments then we can merge

File: crates/noirc_frontend/src/parser/parser.rs
Comment (vezenovm) [reply crates/noirc_frontend/src/parser/parser.rs]: Oops missed those thank you, removed

File: crates/noirc_evaluator/src/ssa/code_gen.rs
Comment (vezenovm) [reply crates/noirc_evaluator/src/ssa/code_gen.rs]: Done

**Review by jfecher — APPROVED** [review]
LGTM

**Pull request merged**
